### PR TITLE
Calendar activity

### PR DESCRIPTION
**Pull request by amitsagtani97**

### CODE REVIEW

File: src/activities/Calender/ActivityInfo.qml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/* GCompris - ActivityInfo.qml
+ *
+ * Copyright (C) 2017 Your Name <yy@zz.org>
```
Comment (petitlapin) [inline src/activities/Calender/ActivityInfo.qml]: fix copyright

File: src/activities/Calender/ActivityInfo.qml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+import GCompris 1.0
+
+ActivityInfo {
+  name: "Calender/Calender.qml"
+  difficulty: 1
+  icon: "Calender/Calender.svg"
+  author: "Your Name &lt;yy@zz.org&gt;"
+  demo: true
+  title: "Calender activity"
```
Comment (petitlapin) [inline src/activities/Calender/ActivityInfo.qml]: qsTr() for translatable texts

File: src/activities/Calender/ActivityInfo.qml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+import GCompris 1.0
+
+ActivityInfo {
+  name: "Calender/Calender.qml"
+  difficulty: 1
+  icon: "Calender/Calender.svg"
+  author: "Your Name &lt;yy@zz.org&gt;"
```
Comment (petitlapin) [inline src/activities/Calender/ActivityInfo.qml]: name

File: src/activities/Calender/ActivityInfo.qml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+import GCompris 1.0
+
+ActivityInfo {
+  name: "Calender/Calender.qml"
```
Comment (petitlapin) [inline src/activities/Calender/ActivityInfo.qml]: the activity should be "calendar", not "calender"

File: src/activities/Calender/ActivityInfo.qml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  demo: true
+  title: "Calender activity"
+  description: ""
+  //intro: "put here in comment the text for the intro voice"
+  goal: ""
+  prerequisite: ""
+  manual: ""
+  credit: ""
+  section: "fun"
+  createdInVersion:8000
```
Comment (petitlapin) [inline src/activities/Calender/ActivityInfo.qml]: should be 9000 (will be the next version)

File: src/activities/Calender/Calender.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
+/* GCompris - Calender.js
+ *
+ * Copyright (C) 2017 YOUR NAME <xx@yy.org>
```
Comment (petitlapin) [inline src/activities/Calender/Calender.js]: copyright

File: src/activities/Calender/Calender.qml [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            gradient: Gradient {
+                GradientStop { position: 0.0; color: "#000" }
+                GradientStop { position: 0.9; color: "#666" }
+                GradientStop { position: 1.0; color: "#AAA" }
+            }
+        }
+
+        GCText {
+            id: questionItem
+            text: "The date is: " + new Date().toLocaleDateString(Qt.locale("de_DE")) + " \n  " +new Date().toLocaleDateString(Qt.locale("en_US"))
```
Comment (petitlapin) [inline src/activities/Calender/Calender.qml]: qsTr() for translatable text. .arg() for arguments
locale should be gotten from ApplicationSettings.locale

File: src/activities/activities.txt [hunk 8]
```diff
@@ -9,15 +9,16 @@ algorithm
 align4
 align4-2players
 alphabet-sequence
-baby_wordprocessor
```
Comment (petitlapin) [inline src/activities/activities.txt]: remove these changes that are not useful

File: src/activities/activities.txt [hunk 9]
```diff
@@ -9,15 +9,16 @@ algorithm
...
 babymatch
 babyshapes
+baby_wordprocessor
 balancebox
 ballcatch
 bargame
 bargame_2players
 braille_alphabets
 braille_fun
+Calender
```
Comment (petitlapin) [inline src/activities/activities.txt]: lowercase calendar

File: src/activities/Calender/Calender.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    if(--currentLevel < 0) {
+        currentLevel = numberOfLevel - 1
+    }
+    initLevel();
+}
+
+function levelConfigurations() {
+    switch(items.bar.level) {
+    case 1:
+        items.calender.navigationBarVisible = false
```
Comment (petitlapin) [inline src/activities/Calender/Calender.js]: wouldn't it be better to have these in the dataset?

**Comment by petitlapin** [comment]
qrc:/gcompris/src/activities/menu/Menu.qml:379:17: QML Image: Cannot open: qrc:/gcompris/src/activities/Calender/Calender.svg

File: src/activities/calendar/ActivityInfo.qml [hunk 11]
```diff
@@ -0,0 +1,35 @@
+/* GCompris - ActivityInfo.qml
+ *
+ * Copyright (C) 2017 Your Name <yy@zz.org>
```
Comment (petitlapin) [inline src/activities/calendar/ActivityInfo.qml]: Copyright

File: src/activities/calendar/Calendar.js [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        items.calendar.navigationBarVisible = false
+        items.calendar.visibleMonth = 2
+        items.calendar.visibleYear = 2018
+        items.calendar.minimumDate = "2018-03-01"
+        items.calendar.maximumDate = "2018-03-31"
+        currentDataSet = dataset[0]
+        currentDataSet = Core.shuffle(currentDataSet)
+        break;
+    case 2:
+        items.calendar.navigationBarVisible = false
```
Comment (petitlapin) [inline src/activities/calendar/Calendar.js]: you can put all of these data in the dataset

File: src/activities/calendar/Calendar.js [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+function checkAnswer() {
+    switch(items.bar.level) {
+    case 1:
+        if(dateSelected.getDay() === correctAnswer) {
+            if(currentDataSet.length > currentSubLevel)
+                items.bonus.good("lion")
+            items.questionDelay.start()
+        }
+        else
+            items.okButtonParticles.burst(20)
```
Comment (petitlapin) [inline src/activities/calendar/Calendar.js]: why having a "positive" burst when the answer is bad?

File: src/activities/calendar/Calendar.js [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        correctAnswer = currentDataSet[currentSubLevel-1]["answer"]
+    }
+}
+
+function checkAnswer() {
+    switch(items.bar.level) {
+    case 1:
+        if(dateSelected.getDay() === correctAnswer) {
+            if(currentDataSet.length > currentSubLevel)
+                items.bonus.good("lion")
```
Comment (petitlapin) [inline src/activities/calendar/Calendar.js]: it would be better to have the good bonus only when finishing the whole level

File: src/activities/calendar/ChoiceTable.qml [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+// Creates a active button for making a table of days of weeks.
+// Set different behaviours and animations on selecting a day from day table here.
+
+import QtQuick 2.6
+import GCompris 1.0
+import QtQuick.Controls 1.4
+import "../../core"
+import "calendar.js" as Activity
+
+Rectangle {
```
[the remark this answers is not in the document]
Comment (amitsagtani97) [reply src/activities/calendar/ChoiceTable.qml]: I checked the animation of missing_letter activity button, the current animation is same as that of "Colour Activity" (on correct guess of duck... rotation and particleStar animations starts... on wrong guess cross sign is displayed), I think the present animation is good.

File: src/activities/activities.txt [hunk 16]
```diff
@@ -1,5 +1,3 @@
-# The list of activities that will be loaded at GCompris start.
-# Keep it sorted
```
Comment (divyam3897) [inline src/activities/activities.txt]: Why you deleted these comments?

File: src/activities/activities.txt [hunk 17]
```diff
@@ -137,3 +136,4 @@ tic_tac_toe_2players
 traffic
 watercycle
 wordsgame
+
```
Comment (divyam3897) [inline src/activities/activities.txt]: Empty line?

**Comment by petitlapin** [comment]
Hi,
it's starting to be really interesting :). Can you apply the following patch (if it is good for you): https://paste.kde.org/pkhx1h3dj
Can you take a look at hiding the left/right navigation button if we are at the limits of the minimum/maximum dates for higher levels (basically, the child shouldn't be able to go before 2017 and after 2019 on these cases)?
Also, it would be great to be able to navigate in the calendar with the keyboard, can you try it?

Johnny

**Comment by petitlapin** [comment]
A few minor changes and it would be good for me to merge it.
I would reword some questions:
* "Select Date .." -> "Select day .."
* "What day of week is on Date 4 of given month" -> "What day of week is the 4th of given month?"
* "Select a Date having Monday between Date 1 and 7 of given month." -> "Select a Monday between days 1 and 7 of given month." (I'm not sure if we keep the uppercase for Monday to emphasize a little or if it's more correct to have a lowercase)
* "Select 2nd Day before 15th Of given month" -> "Select second day before the 15th of given month"

You can also add a new level like: "Find a month starting a Monday and having 31 days" with only one year visible (need to check if multiple answer possible to be sure all choices are valid).

For the min/max date, for now it can go one month before/after the limit, it should stop at the good date (the code only check the year, I think it can be done without the new variables minYearVisible and maxYearVisible).

Can you check if the current code is extensible easily to add another activity "Find the day" based on duration computation: "Find the date 13 days after the 03/05 (03/05 not included)."? I think it is good with the current code and this activity can be done in another PR.

**Comment by Animtim** [comment]
Hi, for me it looks good globally. Just one little thing on the interface: can you please add the day names on the top of each column? (Sun. Mon. Tue. Wed. Thu. Fri. Sat. )

**Comment by amitsagtani97** [comment]
Hi, the following changes are made

- [x] Fixing vertical layout

- [x] Fixing days of week visibility

- [x]  Adding a new level

- [x] Adding animation on correct answer

- [x] Testing if this activity can be extended to add a new activity "Find the day"

File: src/activities/calendar/calendar.js [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+        items.questionItem.text = currentDataSet[currentSubLevel-1]["question"]
+        correctAnswer = currentDataSet[currentSubLevel-1]["answer"]
+    }
+}
+
+function checkAnswer() {
+    // For levels having days of week table visible.
+    if(items.answerChoices.visible) {
+        if(dayOfWeekSelected === correctAnswer["dayOfWeek"]) {
+            items.questionDelay.start()
```
Comment (petitlapin) [inline src/activities/calendar/calendar.js]: code duplication of the if/else

File: src/activities/calendar/Calendar.qml [hunk 19]
```diff
@@ -0,0 +1,459 @@
...
+        Score {
+            id: score
+            fontSize: items.horizontalLayout ? 16 : (answerChoices.visible ? 12 : 8)
+            anchors.top: calendarBox.top
+            anchors.bottom: undefined
+            anchors.left: calendarBox.right
+            anchors.right: undefined
+            anchors.margins: items.horizontalLayout ? 30 : 10
+            property alias winAnimation: winAnmations
+            SequentialAnimation {
```
Comment (petitlapin) [inline src/activities/calendar/Calendar.qml]: if it can be used on other activities, it would be better to have it directly in Score.qml and have a signal to trigger it

File: src/activities/calendar/Calendar.qml [hunk 20]
```diff
@@ -0,0 +1,422 @@
...
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+import QtQuick 2.6
+import GCompris 1.0
+import QtQuick.Controls 1.4
```
Comment (petitlapin) [inline src/activities/calendar/Calendar.qml]: 1.5

File: src/activities/calendar/Calendar.qml [hunk 21]
```diff
@@ -0,0 +1,422 @@
...
+import QtQuick.Controls.Private 1.0
+import QtQuick.Controls.Styles 1.1
```
Comment (petitlapin) [inline src/activities/calendar/Calendar.qml]: 1.4

File: src/activities/calendar/ChoiceTable.qml [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Creates a active button for making a table of days of weeks.
+// Set different behaviours and animations on selecting a day from day table here.
+
+import QtQuick 2.6
+import GCompris 1.0
+import QtQuick.Controls 1.4
```
Comment (petitlapin) [inline src/activities/calendar/ChoiceTable.qml]: 1.5

File: src/activities/calendar/calendar.js [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+    else
+        items.bonus.bad("lion")
+    isCorrectAnswer = false
+}
+
+function checkAnswer() {
+    // For levels having questions based on day of week only.
+    if(mode === "findDayOfWeek") {
+        if(dayOfWeekSelected === correctAnswer["dayOfWeek"]) {
+            isCorrectAnswer = true
```
Comment (petitlapin) [inline src/activities/calendar/calendar.js]: it would be better to have a local variable instead of a global one and pass it as argument in the function

File: src/activities/calendar/calendar.js [hunk 24]
```diff
@@ -0,0 +1,141 @@
+/* GCompris - calendar.js
+ *
+ * Copyright (C) 2017 Amit Sagtani <asagtani06@gmail.com>
+ *
+ * Authors:
+ *   "Amit Sagtani" <asagtani06@gmail.com>
```
Comment (petitlapin) [inline src/activities/calendar/calendar.js]: you don't need the quotes around your name

**Comment by Animtim** [comment]
A few last things:
You can improve the calendar layout to be wider on all display ratio with two small changes in Calendar.qml:
-set bool horizontalLayout to background.width > background.height * 1.5
-in calendarBox, set witdth for horizontal layout to 0.50

Also, small typo in the dataset: it's 22nd, not 22th ;)

**Comment by amitsagtani97** [comment]
@petitlapin @Animtim   All the required changes are made :)

**Comment by petitlapin** [comment]
Merged in https://commits.kde.org/gcompris/b9c0a8905c5688f57605e6d8784e31b25f854d79
Thank you!

Johnny
